### PR TITLE
feat(gatsby-theme-bodiless): Implements delete page functionality (#392)

### PR DESCRIPTION
**Pull request by aeciotr**
## Changes
  - Add delete icon in menu edit page
  - Add functionality delete page

## Test Instructions
  - https://github.com/johnsonandjohnson/Bodiless-JS/issues/392
 
## Related Issues
https://github.com/johnsonandjohnson/Bodiless-JS/issues/392

### CODE REVIEW

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    actualState = -1;
+    return;
+  }
+
+  actualState = -1;
+
+  const { href } = window.location;
+  const hrefArray = href.split('/');
+
+  // Handle paths withtout '/' at the end.
```
Comment (VancheeZze) [inline packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx:151]: Little typo in comment `withtout`.
So here we need an empty string in arrays without trailing slash to work with `splice`, right?
What if we do something like this:
```js
hrefArray[hrefArray.length - 1] !== ''
  ? hrefArray.splice(-1, 1)
  : hrefArray.splice(-2, 1);
```

File: packages/bodiless-backend/src/page.js [hunk 2]
```diff
@@ -94,6 +103,35 @@ class Page {
...
+       */
+      const [, pageRelativeDir] = this.directory.split('/data/pages/');
+      if (!pageRelativeDir) {
+        resolve(`Invalid directory "${this.directory}" to delete`);
+        return;
+      }
+
+      const subdirs = getDirectories(this.directory);
+      if (subdirs.length !== 0) {
+        resolve('The page cannot be deleted it has child pages. To delete this page, first delete or move all child pages, and retry.');
```
[the remark this answers is not in the document]
Comment (aeciotr) [reply packages/bodiless-backend/src/page.js:122]: Yeah, we keep the request successful but letting the client know the action wasn't performed.

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+  Complete,
+  Errored,
+}
+
+type PageStatus = {
+  status: DeletePageState;
+  errorMessage?: string;
+};
+
+type DeletePageProps = PageStatus;
```
[the remark this answers is not in the document]
Comment (aeciotr) [reply packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx]: Ready, it was fixed.

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  if (result.response) {
+    if (result.message !== 'Success' && typeof (result.message) === 'string') {
+      return Promise.reject(new Error(result.message));
+    }
+    return Promise.resolve();
+  }
+  return Promise.reject(new Error('The page cannot be deleted.'));
+};
+
+const DeletePageComp = (props : DeletePageProps) => {
```
[the remark this answers is not in the document]
Comment (aeciotr) [reply packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx]: Ready, it was fixed.

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    ComponentFormLabel,
+    ComponentFormDescription,
+    ComponentFormWarning,
+    ComponentFormTitle,
+    ComponentFormLink,
+  } = defaultUI;
+  const formTitle = 'Delete (this) Page';
+  switch (status) {
+    case DeletePageState.Init: {
+      const CustomComponentFormLabel = flow(
```
[the remark this answers is not in the document]
Comment (aeciotr) [reply packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx]: Ok thanks, fixed with suggestions.

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        removeClasses('bl-text-xs'),
+        addClasses('bl-font-bold bl-text-sm'),
+      )(ComponentFormLabel as ComponentType<StylableProps>);
+      const CustomComponentFormLink = flow(
+        removeClasses('bl-block'),
+        addClasses('bl-italic'),
+      )(ComponentFormLink as ComponentType<StylableProps>);
+      const CustomComponentFormWarning = flow(
```
[the remark this answers is not in the document]
Comment (aeciotr) [reply packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx]: Ok thanks, fixed with suggestions.

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        removeClasses('bl-float-left'),
+      )(ComponentFormWarning);
+      const ui = {
+        ...defaultUI,
+        ComponentFormLabel: CustomComponentFormLabel as ComponentType<HTMLProps<HTMLLabelElement>>,
+        ComponentFormLink: CustomComponentFormLink as ComponentType<HTMLProps<HTMLAnchorElement>>,
+        ComponentFormWarning: CustomComponentFormWarning,
```
[the remark this answers is not in the document]
Comment (aeciotr) [reply packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx]: Ok thanks, fixed with suggestions.

File: packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx (hunk 1 above)
Comment (aeciotr) [reply packages/gatsby-theme-bodiless/src/dist/withDeletePageButton.tsx:134]: It does not work as suggested. The Lint sintaxe broke in rule eslint/no-unused-expressions .

**Review by VancheeZze — APPROVED** [review]

**Review by VancheeZze — DISMISSED** [review]

**Pull request merged**